### PR TITLE
chore: optimise first evaluation by adding worker scope cache (CE)

### PR DESCRIPTION
**Pull request by dvj1988**
Fixes #`Issue Number`  
_or_  
Fixes `Issue URL`
> [!WARNING]  
> _If no issue exists, please create an issue first, and check with the maintainers if the issue is valid._

## Automation

/ok-to-test tags="@tag.All"

### :mag: Cypress test results

> [!TIP]
> 🟢 🟢 🟢 All cypress tests have passed! 🎉 🎉 🎉
> Workflow run: <https://github.com/appsmithorg/appsmith/actions/runs/12295048843>
> Commit: 04b1e859b02282ba9efa96aea25acc9f20098061
> <a href="https://internal.appsmith.com/app/cypress-dashboard/rundetails-65890b3c81d7400d08fa9ee5?branch=master&workflowId=12295048843&attempt=1" target="_blank">Cypress dashboard</a>.
> Tags: `@tag.All`
> Spec:
> <hr>Thu, 12 Dec 2024 12:21:04 UTC

## Communication
Should the DevRel and Marketing teams inform users about this change?
- [ ] Yes
- [ ] No

## Summary by CodeRabbit

- **New Features**
	- Introduced a new property `actionNames` to enhance action tracking within JavaScript action entities.
	- Added a new feature flag `release_evaluation_scope_cache` for improved feature management.
	- Implemented a new function `isPropertyAnEntityAction` to identify action properties in JavaScript entities.
	- Enhanced the `loadAppEntities` method to improve JavaScript library loading processes.
	- Updated the evaluation context initialization process to utilize `getDataTreeContext`.
	- Expanded the `WIDGET_CONFIG_MAP` to include detailed configurations for various widget types.

- **Bug Fixes**
	- Enhanced error handling for unsafe function calls in evaluation logic.
	- Improved error handling and logging for library installation and uninstallation processes.

- **Tests**
	- Expanded test coverage for action bindings and dependencies in the `DataTreeEvaluator`.
	- Updated tests to validate the new `actionNames` property and its integration.
	- Modified tests to ensure correct functionality of the `evaluateSync` function.
	- Added new test cases to assess the behavior of the evaluator with widget updates.

- **Chores**
	- Streamlined imports and initialization of worker instances across various files.

### CODE REVIEW

File: app/client/src/workers/common/DataTreeEvaluator/index.ts [hunk 1]
```diff
@@ -1084,6 +1092,18 @@ export default class DataTreeEvaluator {
     const { isFirstTree, metaWidgets, unevalUpdates } = options;
     let staleMetaIds: string[] = [];
 
+    const triggerBasedDataTreeContext = getDataTreeContext({
```
Comment (dvj1988) [inline app/client/src/workers/common/DataTreeEvaluator/index.ts:1095]: Trigger based eval context is probably not required. Create a new pr and test with only nonTriggerBasedDataTreeContext

File: app/client/src/sagas/EvaluationsSaga.ts [hunk 2]
```diff
@@ -305,6 +294,8 @@ export function* evaluateTreeSaga(
     evalTreeRequestData,
   );
 
+  yield call(waitForFetchEnvironments);
```
[the remark this answers is not in the document]
Comment (rishabhrathod01) [reply app/client/src/sagas/EvaluationsSaga.ts:297]: waiting on this to be moved

File: app/client/src/workers/Evaluation/evaluate.ts [hunk 3]
```diff
@@ -365,15 +366,16 @@ export function setEvalContext({
...
   userScript: string,
   dataTree: DataTree,
   isJSCollection: boolean,
   context?: EvaluateContext,
   // TODO: Fix this the next time the file is edited
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
   evalArguments?: Array<any>,
-  configTree?: ConfigTree,
+  configTree: ConfigTree = {},
+  scopeCache?: EvalContext,
```
Comment (rishabhrathod01) [inline app/client/src/workers/Evaluation/evaluate.ts:378]: Can we name the variable to be able to explain it's usecase. 
```suggestion
  evalContextCache?: EvalContext,
```

**Review by rishabhrathod01 — COMMENTED** [review]
Added comments

File: app/client/src/workers/Evaluation/handlers/jsLibrary.ts [hunk 4]
```diff
@@ -1,4 +1,3 @@
-import { createMessage, customJSLibraryMessages } from "ee/constants/messages";
```
Comment (dvj1988) [inline app/client/src/workers/Evaluation/handlers/jsLibrary.ts:1]: Note to the reviewer: This change was done to reduce the evalWorker chunk size. This change reduces the size by 20KB.

File: app/client/src/workers/common/DataTreeEvaluator/index.ts [hunk 5]
```diff
@@ -1209,6 +1233,13 @@ export default class DataTreeEvaluator {
             set(contextTree, fullPropertyPath, parsedValue);
             set(safeTree, fullPropertyPath, klonaJSON(parsedValue));
 
+            if (
+              WorkerEnv.flags.release_evaluation_scope_cache &&
+              evalContextCache
+            ) {
+              set(evalContextCache, fullPropertyPath, klona(parsedValue));
+            }
```
Comment (vsvamsi1) [inline app/client/src/workers/common/DataTreeEvaluator/index.ts:1241]: Can you use klonaJSON here

Comment (dvj1988) [reply app/client/src/workers/common/DataTreeEvaluator/index.ts]: It was a miss. Changed it to klonaJSON.

**Review by vsvamsi1 — APPROVED** [review]

**Pull request merged**